### PR TITLE
fix(aws): Extend opensearch_service_domains_use_cognito_authentication_for_kibana with SAML

### PR DESCRIPTION
**Pull request by kagahd**
### Context 

The check `opensearch_service_domains_use_cognito_authentication_for_kibana` fails if Amazon OpenSearch has Amazon Cognito authentication for Kibana **not** enabled. However, besides Cognito, AWS recommends to use **SAML** authentication for Kibana which means that Cognito must be disabled. Since the `opensearch_service_domains_use_cognito_authentication_for_kibana` does not test whether **either** Cognito **or** SAML is enabled, it will fail when SAML is enabled.

### Description

Therefore, I renamed the check `opensearch_service_domains_use_cognito_authentication_for_kibana` to `opensearch_service_domains_use_cognito_or_saml_authentication_for_kibana` and extended it in a way that it tests whether **either** Cognito **or** SAML is enabled for Kibana on AWS OpenSearch. If either is enabled, `opensearch_service_domains_use_cognito_or_saml_authentication_for_kibana` will pass.
That's by the way also what the corresponding prowler v2 check `check_extra780` did at its time. See its corresponding [pull request](https://github.com/prowler-cloud/prowler/pull/1291) that had been merged into main at that time. It's a pity that this change hadn't been migrated neither to prowler v3 nor to v4.

Since my company has not yet migrated from prowler v3 to v4, I'm opening this PR for v3 first. Do you want me to open a PR also for prowler v4?

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jfagoagas** [comment]
Hi @kagahd, thanks for your contribution!

We'll review the PR as soon as possible but some things to consider:
- We cannot rename the check since it'll introduce a breaking change in the outputs and unique ids.
- If you can open this PR too to v4 it'd be great.
- We have to include more tests cases to handle this new behaviour, I can help you with that.

**Comment by kagahd** [comment]
Hi @jfagoagas , thanks for your fast reply.
Ok, I rename the check to its former name but then its title would be a bit misleading. Is this ok?
Great to have your help for adding more unit tests!

**Comment by jfagoagas** [comment]
> Hi @jfagoagas , thanks for your fast reply. Ok, I rename the check to its former name but then its title would be a bit misleading. Is this ok? Great to have your help for adding more unit tests!

The title and description can be kept as you did I think. Please use the configured `pre-commit` to format code in each push, refer here to know more about how to configure and use it https://docs.prowler.com/projects/prowler-open-source/en/latest/developer-guide/introduction/#contributing-with-your-code-or-fixes-to-prowler.

**Comment by kagahd** [comment]
I already renamed back the check to its original name before I installed `pre-commit`? Do I need to commit and push again to let `pre-commit` kick in?

**Comment by jfagoagas** [comment]
> I already renamed back the check to its original name before I installed `pre-commit`? Do I need to commit and push again to let `pre-commit` kick in?

I've just made some changes for you to review it but I need push access to your forked branch, could you give me access?

**Comment by kagahd** [comment]
> I've just made some changes for you to review it but I need push access to your forked branch, could you give me access?

I added you.
I also added unit tests. I hope we were not cross-working.

File: tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py [hunk 1]
```diff
@@ -87,7 +88,39 @@ def test_cognito_enabled(self):
...
+
+            check = opensearch_service_domains_use_cognito_authentication_for_kibana()
+            result = check.execute()
+            assert len(result) == 1
+            assert result[0].status == "PASS"
+            assert search(
+                "has either Amazon Cognito authentication or SAML authentication for Kibana enabled",
                 result[0].status_extended,
             )
             assert result[0].resource_id == domain_name
```
Comment (jfagoagas) [inline tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py]:
```suggestion
            assert len(result) == 1
            assert result[0].status == "PASS"
            assert (
                result[0].status_extended
                == f"Opensearch domain {domain_name} has either Amazon Cognito or SAML authentication for Kibana enabled."
            )
            assert result[0].region == AWS_REGION_EU_WEST_1
            assert result[0].resource_id == domain_name
            assert result[0].resource_arn == domain_arn
            assert result[0].resource_tags == []
```

File: tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py [hunk 2]
```diff
@@ -87,7 +88,39 @@ def test_cognito_enabled(self):
...
+        opensearch_client.opensearch_domains = []
+        opensearch_client.opensearch_domains.append(
+            OpenSearchDomain(
+                name=domain_name,
+                region=AWS_REGION_EU_WEST_1,
+                arn=domain_arn,
+                saml_enabled=True,
+            )
+        )
+        opensearch_client.opensearch_domains[0].logging = []
```
Comment (jfagoagas) [inline tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py]:
```suggestion
```

File: tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py [hunk 3]
```diff
@@ -87,7 +88,39 @@ def test_cognito_enabled(self):
             assert len(result) == 1
```
Comment (jfagoagas) [inline tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py:87]: Please change this test assert by
```
assert len(result) == 1
            assert result[0].status == "PASS"
            assert (
                result[0].status_extended
                == f"Opensearch domain {domain_name} has either Amazon Cognito or SAML authentication for Kibana enabled."
            )
            assert result[0].region == AWS_REGION_EU_WEST_1
            assert result[0].resource_id == domain_name
            assert result[0].resource_arn == domain_arn
            assert result[0].resource_tags == []
```

File: tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py [hunk 4]
```diff
@@ -55,7 +56,7 @@ def test_no_cognito_enabled(self):
...
             assert result[0].status == "FAIL"
             assert search(
-                "does not have Amazon Cognito authentication for Kibana enabled",
+                "has neither Amazon Cognito authentication nor SAML authentication for Kibana enabled.",
```
Comment (jfagoagas) [inline tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py]: The same here, change this test assert to
```
assert len(result) == 1
            assert result[0].status == "FAIL"
            assert (
                result[0].status_extended
                == f"Opensearch domain {domain_name} has neither Amazon Cognito nor SAML authentication for Kibana enabled."
            )
            assert result[0].region == AWS_REGION_EU_WEST_1
            assert result[0].resource_id == domain_name
            assert result[0].resource_arn == domain_arn
            assert result[0].resource_tags == []
```

File: prowler/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana.py [hunk 5]
```diff
@@ -14,10 +14,10 @@ def execute(self):
             report.resource_arn = domain.arn
             report.resource_tags = domain.tags
             report.status = "PASS"
-            report.status_extended = f"Opensearch domain {domain.name} has Amazon Cognito authentication for Kibana enabled."
-            if not domain.cognito_options:
+            report.status_extended = f"Opensearch domain {domain.name} has either Amazon Cognito authentication or SAML authentication for Kibana enabled."
```
Comment (jfagoagas) [inline prowler/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana.py]:
```suggestion
            report.status_extended = f"Opensearch domain {domain.name} has either Amazon Cognito or SAML authentication for Kibana enabled."
```

File: prowler/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana.py [hunk 6]
```diff
@@ -14,10 +14,10 @@ def execute(self):
...
+            if not domain.cognito_options and not domain.saml_enabled:
                 report.status = "FAIL"
-                report.status_extended = f"Opensearch domain {domain.name} does not have Amazon Cognito authentication for Kibana enabled."
+                report.status_extended = f"Opensearch domain {domain.name} has neither Amazon Cognito authentication nor SAML authentication for Kibana enabled."
```
Comment (jfagoagas) [inline prowler/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana.py]:
```suggestion
                report.status_extended = f"Opensearch domain {domain.name} has neither Amazon Cognito nor SAML authentication for Kibana enabled."
```

File: prowler/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana.metadata.json [hunk 7]
```diff
@@ -11,8 +11,8 @@
   "ResourceIdTemplate": "arn:partition:service:region:account-id:resource-id",
   "Severity": "high",
   "ResourceType": "AwsOpenSearchDomain",
-  "Description": "Check if Amazon Elasticsearch/Opensearch Service domains has Amazon Cognito authentication for Kibana enabled",
-  "Risk": "Amazon Elasticsearch Service supports Amazon Cognito for Kibana authentication.",
+  "Description": "Check if Amazon Elasticsearch/Opensearch Service domains has Amazon Cognito authentication or SAML authentication for Kibana enabled",
```
Comment (jfagoagas) [inline prowler/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana.metadata.json]:
```suggestion
  "Description": "Check if Amazon Elasticsearch/Opensearch Service domains has Amazon Cognito or SAML authentication for Kibana enabled",
```

File: prowler/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana.metadata.json [hunk 8]
```diff
@@ -1,7 +1,7 @@
 {
   "Provider": "aws",
   "CheckID": "opensearch_service_domains_use_cognito_authentication_for_kibana",
-  "CheckTitle": "Check if Amazon Elasticsearch/Opensearch Service domains has Amazon Cognito authentication for Kibana enabled",
+  "CheckTitle": "Check if Amazon Elasticsearch/Opensearch Service domains has either Amazon Cognito authentication or SAML authentication for Kibana enabled",
```
Comment (jfagoagas) [inline prowler/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana.metadata.json]:
```suggestion
  "CheckTitle": "Check if Amazon Elasticsearch/Opensearch Service domains has either Amazon Cognito or SAML authentication for Kibana enabled",
```

**Review by jfagoagas — CHANGES_REQUESTED** [review]

**Comment by jfagoagas** [comment]
> > I've just made some changes for you to review it but I need push access to your forked branch, could you give me access?
> 
> I added you. I also added unit tests. I hope we were not cross-working.

Great, don't worry it's fine, now I don't need to push anything :D 

Remember to run the `pre-commit` as part of the `git commit` to format files and run linters before pushing.

File: tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py [hunk 9]
```diff
@@ -38,6 +37,7 @@ def test_no_cognito_enabled(self):
                 region=AWS_REGION_EU_WEST_1,
                 arn=domain_arn,
                 cognito_options=False,
+                saml_enabled=False,
             )
         )
         opensearch_client.opensearch_domains[0].logging = []
```
Comment (jfagoagas) [inline tests/providers/aws/services/opensearch/opensearch_service_domains_use_cognito_authentication_for_kibana/opensearch_service_domains_use_cognito_authentication_for_kibana_test.py]:
```suggestion
```

**Review by jfagoagas — APPROVED** [review]
That's a great improvement for the check @kagahd 👏 

Thanks for taking care of it also including new tests cases 🚀 

If you can include this work for the v4 it'd be really appreciated 🙌

**Pull request merged**